crates/oq3_semantics/src/asg.rs

```diff
@@ -562,6 +562,23 @@ impl GateDeclaration {
     }
 }
 
+#[derive(Clone, Debug, PartialEq, Eq, Hash)]
+pub struct MeasureExpression {
+    operand: Box<TExpr>,
+}
+
+impl MeasureExpression {
+    pub fn new(operand: TExpr) -> MeasureExpression {
+        MeasureExpression { operand: Box::new(operand) }
+    }
+
+    // FIXME: type may not be correct here.
+    // This assumes a single qubit is measured.
+    pub fn to_texpr(self) -> TExpr {
+        TExpr::new(Expr::MeasureExpression(self), Type::Bit(IsConst::False))
+    }
```

Comment:
I think the type map should be fairly straightforwards, if the program is semantically correct:
```rust
let out_type = match self.operand.get_type() {
    Type::Qubit => Ok(Type::Bit(IsConst::False)),
    Type::QubitArray(dims) => Ok(Type::BitArray(dims.clone(), IsConst::False)),
    _ => Err(<type error>),
}?;
```